krb5: merge krb5 and libkrb5 with krb5.lib output

de11n opened the pull request:
## Description of changes

krb5 and libkrb5 are two separate derivations that can easily end up
in the same closure. They both provide the same shared libraries and
some packages end up getting both copies. Since both copies come from
the same source, packages often get lucky in this situation and just
use whichever library is found first. Sometimes packages are less
fortunate and end up trying to load both. This has gone largely
unnoticed in Nixpkgs, likely because Kerberos is not widely used
outside of enterprise deployments.

This situation seems to have arisen out of a need to break a cycle
in `fetchurl -> curl -> krb5 -> fetchurl`. The libkrb5 build was able to
avoid depending on bison and libedit, making it easier to break the
cycle.

However, we can break the cycle without resorting to two variants of
krb5. Libedit can be removed with configure flags and byacc can be used
instead of bison, allowing a much smaller build closure that can easily
be resolved when breaking the cycle.

This change also adds a "lib" output to krb5 so that packages depending
on krb5 can still benefit from a smaller runtime closure if they only
need the shared libraries.

cc @trofi @rhelmot @SuperSandro2000 @nessdoor @Ericson2314 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nessdoor commented on pkgs/development/libraries/kerberos/krb5.nix:
```diff
@@ -10,58 +18,50 @@
 # Extra Arguments
-, type ? ""
+, withLdap ? false, openldap
+, withLibedit ? true, libedit
```
Can you provide us with a rationale behind the inclusion of `byacc` and `libedit`?

de11n replied:
Ultimately, the reason we used to have `libOnly` was to avoid `bison` and `libedit` dependencies. Using those makes it much harder to break the cycle `fetchurl -> curl -> krb5 -> fetchurl`. I was able to solve that by...

  1. `libedit` is actually an optional dependency if you just disable it with a `configure` flag, so I made it optional and had `fetchurl` disable it.
  2. `bison` has a large build-time closure, but `byacc` has a tiny one. Thankfully, `krb5` says it can use either one, so I switched to `byacc` to make it very easy to break the cycle.

de11n replied:
I've updated the commit message to include this detail and explain the situation better.

nessdoor commented on pkgs/development/libraries/kerberos/krb5.nix:
```diff
@@ -83,34 +83,32 @@ stdenv.mkDerivation rec {
-    runHook postInstall
+  preInstall = ''
+    mkdir -p "$lib"/{bin,sbin,lib/pkgconfig,share/{et,man/man1}}
   '';
```
So basically the original derivation was forcibly overriding a bunch of standard steps to avoid compiling certain parts that would have resulted in a circular dependency, right?

de11n replied:
Like I said in another comment, I think the original derivation was trying to avoid `libedit` and `bison` and it found a way to do that by just building the shared objects. But we can avoid them differently.

It's also fair to say that the original derivation was optimized for the common case: `libkrb5` is used far more than `krb5`. But, that causes issues as I describe in the PR description.

nessdoor reviewed (commented):
I see no obvious problems, I just need some clarification around certain changes.

Due to the lack of computing power, I cannot currently compile any of these outputs, until the build farm does its job.

The only place of concern that I would like to test is the correct functioning of the LDAP extension, but I am entirely to blame for not having provided an automated test in time (as I had promised a long time ago in https://github.com/NixOS/nixpkgs/pull/286043#issuecomment-1928444665).
This is not a stopper, though, as any malfunctioning can be corrected in a future PR and the amount of people who depend on this functionality seems rather small.

K900 commented:
Does not eval.

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1145,19 +1145,21 @@ with pkgs;
+          byacc = buildPackages.byacc.override { fetchurl = stdenv.fetchurlBoot; };
           keyutils = buildPackages.keyutils.override { fetchurl = stdenv.fetchurlBoot; };
-        };
+        }).lib;
```
Probably needs `dev` too?

de11n replied:
It certainly doesn't *need* it, but I don't think there's a good reason to pick this output anymore. I'll get rid of this `.lib` part.

Ericson2314 commented:
This makes sense to me.

I could imagine still having the binaries in a separate derivation, but importantly that derivation should *not* rebuild the libraries -- the point would be to share libraries between what `fetchurl` needs. The current way doesn't share more, it shares less, so this is strictly better.

de11n commented:
@Ericson2314 I originally wanted to do it that way, but it seemed tricky to build `krb5` in such a way that it wouldn't build the libraries but would reuse them from a different derivation. It doesn't even really support building the libraries separately without manual intervention. We could hack it in, I'm sure, but as you say, this seemed easier and still quite a bit better than what we have now.

Ericson2314 commented:
@de11n Yeah makes sense. I am not sure what level of remaining you have done so far, but we *could* let some `krb5` vs `libkrb5` distinctions in other packages (with the alias) linger in case we find a way of just building the binaries afterwards.

de11n commented:
Would Hydra be ok with an alias in-tree? Or what did you have in mind? If we want that to the eventual end-state, we shouldn't include my `tree-wide` rename and leave everything as `libkrb5`. Anything that actually needs the executables already depends on `krb5`.

de11n commented:
One mess that remains is `krb5-config`. It seems applications using `krb5-config` must largely assume that binaries, shared objects, and headers are all present in `krb5-config --prefix`. If they use `krb5-config --libs`, they will link against the right paths. But if they try to do `$(krb5-config --prefix)/lib`, they won't. Previously, this worked because `libkrb5`'s prefix *only* had libraries. It still would have been wrong for `include`, etc.

This changes makes `krb5-config`'s assumption even more onerous. I fixed one package where this cropped up but haven't found any more. Packages without any tests might end up failing at runtime if they make this assumption.

Ericson2314 commented:
@de11n Yeah I should make clear I am kinda just spit-balling nice to haves --- I always dream of making derivations more fine-grained. I think what you have done as is as fine, and all this is just sort ideal "why do packages be this way" thinking :).

de11n commented:
@Ericson2314 I could, instead of renaming `libkrb5 -> krb5`, do `libkrb5 -> krb5.lib` if you prefer? That is slightly lossy since `krb5.lib.override` doesn't exist, but it would work for the vast majority of cases.

Ericson2314 commented:
Yeah I think the lib stuff will need the `dev` output.

Commit:
krb5: merge krb5 and libkrb5 with krb5.lib output

krb5 and libkrb5 are two separate derivations that can easily end up
in the same closure. They both provide the same shared libraries and
some packages end up getting both copies. Since both copies come from
the same source, packages often get lucky in this situation and just
use whichever library is found first. Sometimes packages are less
fortunate and end up trying to load both. This has gone largely
unnoticed in Nixpkgs, likely because Kerberos is not widely used
outside of enterprise deployments.

This situation seems to have arisen out of a need to break a cycle
in `fetchurl -> curl -> krb5 -> fetchurl`. The libkrb5 build was able to
avoid depending on bison and libedit, making it easier to break the
cycle.

However, we can break the cycle without resorting to two variants of
krb5. Libedit can be removed with configure flags and byacc can be used
instead of bison, allowing a much smaller build closure that can easily
be resolved when breaking the cycle.

This change also adds a "lib" output to krb5 so that packages depending
on krb5 can still benefit from a smaller runtime closure if they only
need the shared libraries.

A future change will include a tree-wide refactor to switch uses of
libkrb5 to krb5.

de11n commented:
FWIW, we've built a large portion of Nixpkgs with these changes and have had no issues so far.

Ericson2314 commented:
@de11n Some conflicts, but otherwise looks good, especially if we skip the treewise rename commit for now.

I would maybe shoot an email to https://mailman.mit.edu/mailman/listinfo/krbdev linking this PR and asking whether they have any insight into building the binaries against preexisting libraries. Just so the next time anyone encounters this stuff there is just the best possible paper trail --- not to block this PR on a response.

de11n commented:
@Ericson2314 I'd happily remove the tree-wide commit, but I heard from someone else that Nixpkgs would not be happy with the use of an alias in-tree. What do you recommend?

de11n commented:
I've resolved the merge conflicts.

Regarding the `libkrb5` -> `krb5` rename: Packages semantically opting into libraries only via `libkrb5` is handy. One way we could preserve that would be to have `libkrb5` be a "dev"-only package that only contains modified `.pc` files which only remember paths to `lib` and `dev`. This would have the downside that expressions like `${libkrb5}/lib` wouldn't work.

It's almost like we want a way to propagate `disallowedReferences` to help temper the magic of automatic run-time dependency discovery. We could also do something like that in the form of a setup hook.

de11n commented:
@Ericson2314 I dropped the tree-wide commit and added an "alias" `libkrb5 = krb5;` in `all-packages.nix` with a `TODO` for what you suggested.

SuperSandro2000 commented on pkgs/development/libraries/kerberos/krb5.nix:
```diff
@@ -133,3 +131,4 @@ stdenv.mkDerivation rec {
     };
   };
 }
+
```
```suggestion
```

de11n replied:
Done

de11n commented:
@Ericson2314 I've fixed `krb5-config --prefix` to point to `krb5.lib` instead of `krb5` (with binaries). This allowed me to *drop* a commit that fixed `perl538Packages.AuthenKrb5Admin`. I moved the commit for `python3Packages.k5test` to #335186 because it doesn't need to be in this PR.

This makes `krb5.dev` more-or-less identical to what `libkrb5.dev` is currently in `master`.

Commit:
openssh: use krb5 dev output

Commit:
libgssglue: use krb5 lib output

Commit:
dotnet: use krb5 lib output
